service/storage_service.cc

```diff
@@ -6153,6 +6189,195 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
+                        .reason = format("The node didn't finish the previous attempt to join group 0, "
+                                "but it was restarted after revoking support for some features that were "
+                                "initially supported ({}). This is not allowed. Please reconfigure the node "
+                                "to support those features or clear the data directory, then restart."),
+                    };
+                    co_return result;
+                }
+                result.result = join_node_request_result::ok {};
+                co_return result;
+            }
+            case node_state::left:
+                result.result = join_node_request_result::rejected{
+                    .reason = "The node has already been removed from the cluster",
+                };
+                co_return result;
+            default:
+                result.result = join_node_request_result::accepted {};
+                co_return result;
+            }
+        }
+
+        if (params.replaced_id.has_value()) {
+            if (!_topology_state_machine._topology.normal_nodes.contains(*params.replaced_id)) {
+                result.result = join_node_request_result::rejected {
+                    .reason = ::format("Cannot replace node {} because it is not in the 'normal' state", *params.replaced_id),
+                };
+                co_return result;
+            }
+        }
+
+        auto mutation = build_mutation_from_join_params(params, *guard);
+
+        topology_change change{{std::move(mutation)}};
+        group0_command g0_cmd = _group0->client().prepare_command(std::move(change), *guard,
+                format("raft topology: placing join request for {}", params.host_id));
+        try {
+            co_await _group0->client().add_entry(std::move(g0_cmd), std::move(*guard), &_abort_source);
+            break;
+        } catch (group0_concurrent_modification&) {
+            slogger.info("raft topology: join_node_request: concurrent operation is detected, retrying.");
+        }
+    }
+
+    slogger.info("raft topology: placed join request for {}", params.host_id);
+
+    // Success
+    result.result = join_node_request_result::ok {};
+    co_return result;
+}
+
+future<join_node_response_result> storage_service::join_node_response_handler(join_node_response_params params) {
+    assert(this_shard_id() == 0);
+
+    // Serialize handling the responses
+    auto lock = co_await get_units(_join_node_response_handler_mutex, 1);
+
+    // Wait until the join_node_request is sent.
+    co_await _join_node_request_done.get_shared_future(_abort_source);
+
+    if (_join_node_result.available()) {
+        // We already handled this RPC. Return immediately for idempotence.
+        slogger.info("raft topology: the node got join_node_response RPC for the second time, ignoring");
+        join_node_response_result result;
+        co_return result;
```

Comment:
nit: Why split to two lines instead if `co_return join_node_response_result{}`? Below as well.

Reply:
No good reason. Replaced them with `co_return join_node_response_result{}`